src/linear_algebra/eigenspace.lean

```diff
@@ -0,0 +1,231 @@
+/-
+Copyright (c) 2020 Alexander Bentkamp. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alexander Bentkamp.
+-/
+
+import field_theory.algebraic_closure
+import linear_algebra.finsupp
+
+/-!
+# Eigenvectors and eigenvalues
+
+This file defines eigenspaces and eigenvalues.
+
+An eigenspace of a linear map `f` for a scalar `μ` is the kernel of the map `(f - μ • id)`. The
+nonzero elements of an eigenspace are eigenvectors `x`. They have the property `f x = μ • x`. If
+there are eigenvectors for a scalar `μ`, the scalar `μ` is called an eigenvalue.
+
+There is no consensus in the literature whether `0` is an eigenvector. Our definition of
+`eigenvector` permits only nonzero vectors. For an eigenvector `x` that may also be `0`, we write
+`x ∈ eigenspace f μ`.
+
+## Notations
+
+The expression `algebra_map K (End K V)` appears very often, which is why we use `am` as a local
+notation for it.
+
+## References
+
+* [Sheldon Axler, *Down with determinants!*,
+  https://www.maa.org/sites/default/files/pdf/awards/Axler-Ford-1996.pdf][axler1996]
+* https://en.wikipedia.org/wiki/Eigenvalues_and_eigenvectors
+
+## Tags
+
+eigenspace, eigenvector, eigenvalue, eigen
+-/
+
+universes u v w
+
+open vector_space principal_ideal_ring polynomial finite_dimensional
+
+variables {K : Type v} {V : Type w} [add_comm_group V]
+
+local notation `am` := algebra_map K (module.End K V)
+
+/-- The subspace `eigenspace f μ` for a linear map `f` and a scalar `μ` consists of all vectors `x`
+    such that `f x = μ • x`. -/
+def eigenspace [comm_ring K] [module K V] (f : module.End K V) (μ : K) : submodule K V :=
+(f - am μ).ker
+
+/-- A nonzero element of an eigenspace is an eigenvector. -/
+def eigenvector [comm_ring K] [module K V] (f : module.End K V) (μ : K) (x : V) : Prop :=
+x ≠ 0 ∧ x ∈ eigenspace f μ
+
+/-- A scalar `μ` is an eigenvalue for a linear map `f` if there are nonzero vectors `x`
+    such that `f x = μ • x`. -/
+def eigenvalue [comm_ring K] [module K V] (f : module.End K V) (a : K) : Prop :=
```

Comment:
This is yet another instance of a discussion we've had in many places, but: would this be better called `is_eigenvalue`?

Reply:
I think I would prefer `is_eigenvalue`.

Reply:
What about `module.End.has_eigenvalue` so that we can use dot notation like this: `f.has_eigenvalue a`?